cocos/platform/android/java/src/org/cocos2dx/lib/Cocos2dxDownloader.java

```diff
@@ -1,319 +1,300 @@
-                    if (null != task.handle) {
-                        task.handle.cancel(true);
-                    }
-                }
-            }
-        });
+    private class DataTaskHandler extends BinaryHttpResponseHandler {
+	int _id;
+	private long _lastBytesWritten;
+
+	void LogD(String msg) {
+	    android.util.Log.d("Cocos2dxDownloader", msg);
+	}
+
+	public DataTaskHandler(int id) {
+	    super(new String[]{".*"});
+	    _id = id;
+	    _lastBytesWritten = 0;
+	}
+
+	@Override
+	public void onProgress(long bytesWritten, long totalSize) {
+	    //LogD("onProgress(bytesWritten:" + bytesWritten + " totalSize:" + totalSize);
+	    long dlBytes = bytesWritten - _lastBytesWritten;
+	    long dlNow = bytesWritten;
+	    long dlTotal = totalSize;
+	    Cocos2dxDownloader.this.onProgress(_id, dlBytes, dlNow, dlTotal);
+	    _lastBytesWritten = bytesWritten;
+	}
+
+	@Override
+	public void onStart() {
+	    Cocos2dxDownloader.this.onStart(_id);
+	}
+
+	@Override
+	public void onFailure(int i, Header[] headers, byte[] errorResponse, Throwable throwable) {
+	    LogD("onFailure(i:" + i + " headers:" + headers + " throwable:" + throwable);
+	    String errStr = null;
+	    if (null != throwable) {
+		errStr = throwable.toString();
+	    }
+	    Cocos2dxDownloader.this.onFinish(_id, i, errStr, null);
+	}
+
+	@Override
+	public void onSuccess(int i, Header[] headers, byte[] binaryData) {
+	    LogD("onSuccess(i:" + i + " headers:" + headers);
+	    Cocos2dxDownloader.this.onFinish(_id, 0, null, binaryData);
+	}
     }
 
-    public Runnable dequeue() {
-        if (!_taskQueue.isEmpty() && _taskQueue.element() == null) {
-            _taskQueue.remove();
-        }
-        if (!_taskQueue.isEmpty()) {
-            return _taskQueue.remove();
-        }
-        return null;
+    private class FileTaskHandler extends FileAsyncHttpResponseHandler {
+	int _id;
+	File _finalFile;
+
+	private long _initFileLen;
+	private long _lastBytesWritten;
+
+	void LogD(String msg) {
+	    android.util.Log.d("Cocos2dxDownloader", msg);
+	}
+
+	public FileTaskHandler(int id, File temp, File finalFile) {
+	    super(temp, true);
+	    _finalFile = finalFile;
+	    _id = id;
+	    _initFileLen = getTargetFile().length();
+	    _lastBytesWritten = 0;
+	}
+
+	@Override
+	public void onProgress(long bytesWritten, long totalSize) {
+	    //LogD("onProgress(bytesWritten:" + bytesWritten + " totalSize:" + totalSize);
+	    long dlBytes = bytesWritten - _lastBytesWritten;
+	    long dlNow = bytesWritten + _initFileLen;
+	    long dlTotal = totalSize + _initFileLen;
+	    Cocos2dxDownloader.this.onProgress(_id, dlBytes, dlNow, dlTotal);
+	    _lastBytesWritten = bytesWritten;
+	}
+
+	@Override
+	public void onStart() {
+	    Cocos2dxDownloader.this.onStart(_id);
+	}
+
+	@Override
+	public void onFinish() {
+	    // onFinish called after onSuccess/onFailure
+	    Runnable taskRunnable = Cocos2dxDownloader.this._taskQueue.poll();
```

Comment:
Also this refactor breaks the original logic. Since the first N tasks should push a null into the _taskQueue, once the first task is finished, it should pop out all the first N nulls in order to start the next non-null task runnable, otherwise when the first N tasks are done, only N nulls are popped and the (N+1)th task is left in the queue untouched.